src/sidebar/services/auth.ts

```diff
@@ -0,0 +1,335 @@
+      // tests.
+      if (Date.now() > token.expiresAt) {
+        /* istanbul ignore next */
+        throw new Error('Refreshed token expired in the past');
+      }
+    } catch {
+      // If refreshing the token fails, the user is simply logged out.
+      return null;
+    }
+
+    if (options.persist) {
+      this._saveToken(token);
+    }
+
+    return token;
+  }
+
+  /**
+   * Retrieve an access token for use with the API
+   */
+  async getAccessToken(): Promise<string | null> {
+    if (!this._tokenInfoPromise) {
+      // No access token is set: determine how to get one. This will depend on
+      // which type of login is being used
+      const grantToken = this._getGrantToken();
+      if (grantToken !== undefined) {
+        // The user is logged in through a publisher/embedder's site and
+        // a grant token has been included in service configuration
+        this._tokenInfoPromise = this._exchangeGrantToken(grantToken);
+      } else if (this._authCode) {
+        // User has authorized through pop-up window and we have an
+        // authorization code we can exchange for an access token
+        this._tokenInfoPromise = this._exchangeAuthCodeForToken(this._authCode);
+        this._authCode = null; // Consume the single-use auth code
+      } else {
+        // Attempt to retrieve stored token from previous session
+        this._tokenInfoPromise = Promise.resolve(this._loadToken());
+      }
+    }
+
+    // Wait for the token to resolve. Ensure that it is valid and that
+    // it wasn't invalidated while it was being fetched. Refresh if needed.
+
+    const origToken = this._tokenInfoPromise;
+    const token = await this._tokenInfoPromise;
+
+    if (!token) {
+      // No token available. User will need to log in.
+      return null;
+    }
+
+    if (origToken !== this._tokenInfoPromise) {
+      // The token source was changed while waiting for the token to be fetched.
+      // This can happen for various reasons. We'll just need to try again.
+      return this.getAccessToken();
+    }
+
+    if (Date.now() > token.expiresAt) {
+      // Token has expired, so we need to fetch a new one.
+      const grantToken = this._getGrantToken();
+      this._tokenInfoPromise = this._refreshAccessToken(token.refreshToken, {
+        // Only persist tokens if automatic login is not being used.
+        persist: typeof grantToken === 'undefined',
+      });
+      return this.getAccessToken();
+    }
+
+    return token.accessToken;
+  }
+
+  /**
+   * Log in to the service using OAuth.
+   *
+   * Authorize through OAuthClient (this shows a pop-up window to the user).
+   * Store the resulting authorization code to exchange for an access token in
+   * the next API call ( @see getAccessToken )
+   */
+  async login() {
+    // Any async steps before the call to `client.authorize` must complete
+    // in less than ~1 second, otherwise the browser's popup blocker may block
```

Comment:
I think the `~1 second` part is out of date. I believe the max delay is now on the order of a few seconds. What matters is that opening a popup must be done within a short window of time after a user gesture (eg. clicking on login button), so we should avoid potentially long-running async steps before the `client.authorize` call.

Reply:
@robertknight Do you know where I might find more information on this? I'm not having Google luck today with combinations of "OAuth" and "popup [blocker]". I'm presuming this is related to...maybe...how long ago an associated user-initiated action happened?

Reply:
> Do you know where I might find more information on this?

When I wrote that comment I think it was an implementation detail of browser engines. Nowadays it is documented deep in the bowls of the HTML specification: https://html.spec.whatwg.org/multipage/interaction.html#transient-activation-duration